pkg/streamingpromql/operators/scalar_scalar_binary_operation.go

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package operators
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/compat"
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+type ScalarScalarBinaryOperation struct {
+	Left                     types.ScalarOperator
+	Right                    types.ScalarOperator
+	Op                       parser.ItemType
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+
+	opFunc             binaryOperationFunc
+	expressionPosition posrange.PositionRange
+}
+
+var _ types.ScalarOperator = &ScalarScalarBinaryOperation{}
+
+func NewScalarScalarBinaryOperation(
+	left, right types.ScalarOperator,
+	op parser.ItemType,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	expressionPosition posrange.PositionRange,
+) (*ScalarScalarBinaryOperation, error) {
+	f := arithmeticOperationFuncs[op]
+	if f == nil {
+		return nil, compat.NewNotSupportedError(fmt.Sprintf("binary expression with '%s'", op))
+	}
+
+	return &ScalarScalarBinaryOperation{
+		Left:                     left,
+		Right:                    right,
+		Op:                       op,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+
+		opFunc:             f,
+		expressionPosition: expressionPosition,
+	}, nil
+}
+
+func (s *ScalarScalarBinaryOperation) GetValues(ctx context.Context) (types.ScalarData, error) {
+	leftValues, err := s.Left.GetValues(ctx)
+	if err != nil {
+		return types.ScalarData{}, err
+	}
+
+	rightValues, err := s.Right.GetValues(ctx)
+	if err != nil {
+		return types.ScalarData{}, err
+	}
+
+	// Binary operations between two scalars always produce a float value, as only arithmetic operators or comparison
+	// operators (with the bool keyword) are supported between two scalars.
+	//
+	// Furthermore, scalar values always have a value at each step.
+	//
+	// So we can just compute the result of each pairwise operation without examining the timestamps of each sample.
+	//
+	// We store the result in the slice from the left operator, and return the right operator's slice once we're done.
```

Comment:
I think for safety we should have a check such as `len(leftValues) != len(rightValues)` and raise an error

Reply:
We rely on this fact in other places, and don't have checks there either (eg. scalar / vector binary operations). 

If the lengths are different, the code will panic, which seems like a reasonable thing to do for a bug where an invariant has been broken.

Reply:
It only panics if the right side is shorter than the left. I don't think the overhead here is very high, what do you think about it adding it to the other scalar operators too?

Reply:
I'm not sure this makes sense as a precedent to set: we don't do similar validation-type checks in other situations (eg. checking that the data points from an instant vector operator are within the query time range, or align to the requested step), and if those situations were to occur, we'd have similarly undefined behaviour.

Reply:
Hmm, okay. I still think it's reasonable, but I don't hold it strongly. LGTM :-)